spec.md

```diff
@@ -179,6 +218,49 @@ physically present in the final circuit. Refer to the description of the
 instance statement for details on how to instantiate a module
 ([@sec:instances]).
 
+### Public Modules
+
+A public module is a module that is marked with the `public`{.firrtl} keyword.
+A public module has a defined ABI---knowing the FIRRTL representation is
+sufficient to know its compiled representation.  This means that a public module
+is available external to this circuit.
+
+The example below shows a public module `Foo`:
+
+``` firrtl
+public module Foo:
+  input a: UInt<1>
+  output b: UInt<1>
+
+  connect b, a
+```
+
+A public module has a number of restrictions:
+
+1. A public module may have no ports of uninferred width.
+1. A public module may have no ports of abstract reset type.
+1. A public module may have no ports of input probe type.
+1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.
```

Comment:
If we clarify the difference, that would help yes, and related issue of whether either is expected to actually force the ports themselves necessarily or only have in some sense equivalent behavior.

This restriction (currently in the spec, just described here now) is to ensure a FIRRTL compiler can emit what's needed on /both/ sides of a module with forced ports, to give room for bounce wires or whatever, since "may be forced" is not part of the signature/ABI.

Emitting ports (and other signals) as variables would mostly eliminate the need for this restriction, as you mention, and we should clarify the difference between rwprobe'ing instance results vs the ports from inside the module so that on, e.g., extmodule's that may not have ports as variables, there's no surprises.

Sound good?

Reply:
What constitutes a "use"?

Is there an intuition for why this rule is in place?